### ISSUE TITLE
In-place modification with double slicing (using boolean mask)

### ISSUE THREAD

**Issue by fabienpesquerel** [pytorch/pytorch#20201]
## 🐛 Bug

Cannot do in-place modification of a tensor with a double slicing using `torch.uint8` (boolean mask) tensor.

## To Reproduce

```
import torch

a = torch.tensor([[1., 2.], [13., 4.], [8., 14.]])

# Creating a mask for those elements of a that are greater than 10
b = a > 10
print("b is: ", b)
print("a[b] is: ", a[b])

# Creating a mask for those elements of a[b] (i.e. [13., 14.]) that are greater than 13.5
c = a[b] > 13.5
print("c is: ", c)
print("a[b][c] is: ", a[b][c])

# HERE IS THE BUG
a[b][c] = torch.tensor([17.], dtype=a.dtype)
# a[b][c] = torch.tensor([17.])
# a[b][c] = 17.

print("a is: ", a)
print("a[b] is: ", a[b])

```

No error is reported when running this code.

## Expected behavior

I think that it is expected that, no matter which of the three ways I described to do the in-place modification of the tensor, `a` should have been modified while it is not. If I use a single mask, then everything works just fine.
If I try something else such as `a[0][0] = 17.` or `a[0, :][0] = 17.` then everything is working (but notice that the slicing is not done with a mask (or `torch.unit8` tensor) in these two cases.

One way to work around this issue is using another tensor:

```
k = a[b]
print("k is: ", k)
k[c] = 42

# k as been modified
print("k is: ", k)

a[b] = k

# a has been modified
print("a is: ", a)
print("a[b] is: ",  a[b])
```

## Environment

 - PyTorch Version: 1.0.1
 - OS: Ubuntu 18.04
 - How I installed PyTorch: `conda install pytorch cudatoolkit=9.0 -c pytorch`
 - Python version: 3.7

**Comment by fmassa** [comment]
This behavior is expected and is the same as in numpy.

There are a few subtleties in here:
- indexing with a boolean tensor always returns a new tensor, and not a view to the existing tensor
- this means that operations on the result of boolean indexing do not modify the indexed tensor
```python
a = torch.randn(5)
b = a[a > 0]
b.fill_(0)  # doesn't modify `a`
# BUT
a[a > 0] = 0  # changes `a`
```
Now, why does `a[a > 0] = 0` modify `a`? Because this dispatches to `__setitem__` internally, which calls the appropriate masking functions.

Now, when you do `a[a > 0.5][a < -0.5] = 0`, what this is in fact doing is that it is first creating a new (temporary) tensor `a[a > 0.5]`, and then on this temporary tensor it applies `t[a < -0.5] = 0` by calling into `__setitem__`. Which means that the result of the operation gets discarded with the temporary tensor.

Now you can say: but why does `x[0][1] = 5` works then? Because `x[0]` returns a view to the underlying tensor, which shares storage.

```python
a = torch.rand(5, 5)
b = a[0]
b.fill_(0)
print(a)  # modified!
```

If you want to compose together boolean indexing, you should first compose the boolean mask (using `&` and `|`), and then apply the indexing. Also, using `torch.where` can be helpful here.

---

Given that this is not a bug and the behavior is expected and follow numpy semantics, I'm closing this issue, but please feel free to reopen it if you feel otherwise.

**Comment by fabienpesquerel** [comment]
Thanks you for this quick and nice answer. I am happy with your explanation although I find this behavior confusing (at least in the first place) but I am sure that there are good reason for this.

The problem was that I was having a high level view of the situation (as often when writing with python).

Thanks again !
